File: js/modules/racks/racks.js
/*
 * Copyright (C) 2015 EDF SA
 *
 * This file is part of slurm-web.
 *
 * slurm-web is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * slurm-web is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with slurm-web.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

define([
  'jquery',
  'async',
  'handlebars',
  'text!../../js/modules/racks/racks.hbs',
  'token-utils',
  '2d-draw',
  '2d-legend-draw'
], function ($, async, Handlebars, template, tokenUtils, D2Draw, d2LegendDraw) {
  template = Handlebars.compile(template);
  var draw = new D2Draw();

  return function (config) {
    this.slurmNodes = null;
    this.interval = null;
    this.config = draw.getConfig();

    this.init = function () {
      var self = this;

      async.parallel({
        nodes: function (callback) {
          var options = {
            type: 'POST',
            dataType: 'json',
            headers: {
              'Accept': 'application/json',
              'Content-Type': 'application/json'
            },
            data: JSON.stringify({
              token: tokenUtils.getToken(config.cluster)
            })
          };

          $.ajax(config.cluster.api.url + config.cluster.api.path + '/nodes', options)
            .success(function (data) {
              callback(null, data)
            })
            .error(function (callback) {
              callback(true, null);
            });
        },
        racks: function (callback) {
          var options = {
            type: 'POST',
            dataType: 'json',
            headers: {
              'Accept': 'application/json',
              'Content-Type': 'application/json'
            },
            data: JSON.stringify({
              token: tokenUtils.getToken(config.cluster)
            })
          };

          $.ajax(config.cluster.api.url + config.cluster.api.path + '/racks', options)
            .success(function (data) {
              callback(null, data);
            })
            .error(function () {
              callback(true, null);
            });
        }
      }, function (err, result) {
        self.slurmNodes = result.nodes;

        var racks = result.racks.racks;
        if (racks instanceof Array) {
          var parsed = {};
          var i;
          var rack;
          for (i in racks) {
            if (racks.hasOwnProperty(i)) {
              for (rack in racks[i]) {
                if (racks[i].hasOwnProperty(rack)) {
                  parsed[rack] = racks[i][rack];
                }
              }
            }
          }
          racks = parsed;
        }

        var context = {
          config: self.config,
          racks: racks
        };

        $('#main').append(template(context));
        $("canvas[id^='cv_rackmap_']").parent('.canvas-container').css('width', self.config.CANVASWIDTH);

        $.each(racks, function (idRack, rack) {
          $('#cv_rackmap_' + idRack).on('mousemove', function (e) {
            e.stopPropagation();
            var offset = $(this).offset();

            $(document).trigger('canvas-mousemove', { rack: idRack, x: (e.pageX - offset.left), y: (e.pageY - offset.top) });
          });

          draw.drawRack(rack);
          $.each(rack.nodes, function (idRackNode, rackNode) {
            draw.drawNode(rack, rackNode, self.slurmNodes[rackNode.name]);
          });
        });

        d2LegendDraw.drawLegend('racks');
      });
    };

    this.refresh = function () {
      var self = this;

      this.interval = setInterval(function () {
        $('#racks').remove();
        self.init();
      }, config.REFRESH);
    };

    this.destroy = function () {
      if (this.interval) {
        clearInterval(this.interval);
      }

      draw.clearNodesHoverIntersections();

      $("canvas[id^='cv_rackmap_']").off('mousemove');
      $('#racks').remove();
    };

    return this;
  };
});

File: src/slurmweb/tests/mocks/racks.py
#!usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015 EDF SA
#
# This file is part of slurm-web.
#
# slurm-web is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# slurm-web is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with slurm-web.  If not, see <http://www.gnu.org/licenses/>.

from xml.dom.minidom import DOMImplementation

class MockRacksXML(object):

    def __init__(self):
        """
          <?xml version="1.0" encoding="UTF-8"?>
          <!DOCTYPE rackmap SYSTEM "/usr/share/slurm-web/restapi/schema/dtd/racks.dtd">
          <rackmap>
            <nodetypes/>
            <racks posx="0" posy="0" width="10" depth="10"/>
          </rackmap>
        """

        self.imp = DOMImplementation()
        self.doctype = self.imp.createDocumentType(
            qualifiedName='rackmap',
            publicId='',
            systemId='/usr/share/slurm-web/restapi/schema/dtd/racks.dtd',
        )
        self.doc = self.imp.createDocument(None, 'rackmap', self.doctype)

        #self.root = self.doc.createElement('rackmap')
        #self.doc.appendChild(self.root)

        self.root = self.doc.documentElement

        self.nodetypes = self.doc.createElement('nodetypes')
        self.root.appendChild(self.nodetypes)
        self.racks = self.doc.createElement('racks')
        self.racks.setAttribute('posx', '0')
        self.racks.setAttribute('posy', '0')
        self.racks.setAttribute('width', '10')
        self.racks.setAttribute('depth', '10')
        self.root.appendChild(self.racks)

    def add_nodetype(self, name, model, height, width):
        """
           <nodetype id="m32x4321"
              model="Vendor A 32 x4321"
              height="1"
              width="0.5"
              />
        """
        nodetype = self.doc.createElement('nodetype')

        nodetype.setAttribute("id", name)
        nodetype.setAttribute("model", model)
        nodetype.setAttribute("height", height)
        nodetype.setAttribute("width", width)

        self.nodetypes.appendChild(nodetype)

    def add_racksrow(self, posx):
        """
          <racksrow posx="0"/>
          Return the <racksrow/> element.
        """
        racksrow = self.doc.createElement('racksrow')
        racksrow.setAttribute('posx', posx)
        self.racks.appendChild(racksrow)
        return racksrow

    def add_rack(self, row, name, posy):
        """
          <rack id="A1" posy="0">
            <nodes/>
          </rack>
          Return <nodes/> element.
        """
        rack = self.doc.createElement('rack')
        rack.setAttribute('id', name)
        rack.setAttribute('posy', posy)
        row.appendChild(rack)
        nodes = self.doc.createElement('nodes')
        rack.appendChild(nodes)
        return nodes

    def add_nodeset(self, nodes, nodelist, nodetype, posx, posy, draw=None):
        """
          <nodeset id="cn[505-518]"
                   type="m32x4321"
                   posx="0"
                   posy="2"
                   draw="down" />
        """
        nodeset = self.doc.createElement('nodeset')
        nodeset.setAttribute('id', nodelist)
        nodeset.setAttribute('type', nodetype)
        nodeset.setAttribute('posx', posx)
        nodeset.setAttribute('posy', posy)
        if draw is not None:
            nodeset.setAttribute('draw', draw)
        nodes.appendChild(nodeset)
